fix(prepared-report): Prepared Report Enhancements

adityahase opened the pull request:
This fixes a few issues with prepared reports.

#### 1. CSV Serialization
CSV serialization causes information loss for erpnext reports like P&L Statement,
rendering such reports unusable.


**Note**: This will cause all previously generated reports to break, that should be fixed.

| Before        | After         |
| ------------- |:-------------:|
| ![before](https://user-images.githubusercontent.com/8528887/48612113-0870ed80-e9ae-11e8-8de2-36d0937d04fa.png)| ![after](https://user-images.githubusercontent.com/8528887/48612264-83d29f00-e9ae-11e8-8a71-db1573f40c82.png) |

#### 2. Permission Issues

Only Prepared Reports owned by the user is being shown.

**To Do**
- [ ] Permissions on Prepared Reports.
- [ ] Refactor how this is being done.

#### 3. Filter Issues

Filters and data didn't match since data from most recently generated report was being shown.

#### 4. Space Issues
Stored files are gzip compressed to reduce space requirement.

**To Do**
- [ ] Scheduled job to delete old prepared reports.

#### 5. UX Improvements
![pr](https://user-images.githubusercontent.com/8528887/49590332-a21a3200-f991-11e8-91b8-05cfe5bd778b.gif)

**Note**: These changes aren't backwards compatible. Provided patch deletes existing Prepared Reports and associated files.

adityahase committed:
fix(prepared-report): Serialize Prepared Report data to JSON

CSV serialization causes information loss for erpnext reports like P&L Statement,
rendering such reports unusable.

Also stored files are gzip compressed to reduce space requirement

Note: This will cause all previously generated reports to break, this should be fixed

rohitwaghchaure commented:
@adityahase For existing prepared data, please write a patch to make JSON files against the CSV

rmehta commented:
Please add tests and fix failing patch

adityahase committed:
fix(prepared-report): Delete all prepared reports

Changes made to Prepared Report aren't backwards compatible.
Let's get rid of existing prepared reports and files associated with them.

adityahase commented:
@nabinhait @rohitwaghchaure @rmehta I've added a patch to remove existing prepared reports.

adityahase committed:
refactor(prepared-report): Do not serialize JSON serialized filters

Prepared report filters were being JSON serialized twice.
There isn't any apparent need to do this. So don't do it.

adityahase committed:
fix(prepared-report): Show data from report matching filters and owner

This fixes permission issues since users can only see the
report that they themselves have generated.

Also report filters and data won't mismatch.

adityahase committed:
fix(prepared-report): Python 2 compatibility fix

gzip.compress and gzip.decompress were introduced in Python 3.2

Added gzip_compress and gzip_decompress to frappe.utils to add this functionality
to both Python 2 and 3

rmehta commented on frappe/desk/query_report.py:
```diff
@@ -102,7 +103,9 @@ def background_enqueue_run(report_name, filters=None, user=None):
 			"doctype": "Prepared Report",
 			"report_name": report_name,
-			"filters": json.dumps(filters),
+			# This looks like an insanity but, without this it'd be very hard to find Prepared Reports matvhing given condition
```
typo "matching"

adityahase committed:
fix(prepared-report): Disable filters when showing a Prepared Report doc

adityahase committed:
fix(prepared-report): Uncompress Prepared Report data files before downloading

adityahase committed:
fix(prepared-report): Set prepared_report_name query param in Link to query report

Since visiting the query report URL will show most recently generated report
that matches given filters, Explicitly mention which Prepared Report is going
to be rendered.

adityahase committed:
fix(prepared-report): Fix typo in a comment

netchampfaris committed:
style: Linting fixes

adityahase committed:
fix(prepared-report): Do not lose the first row in Prepared Report data

netchampfaris committed:
fix: Use frappe.as_json instead of json.dumps

adityahase committed:
fix(prepared-report): UX Improvements

 - Update primary action according to the Report's status

 - Automatically refresh report view if report generation is finished

 - Correctly translate message strings

Co-authored-by: Faris Ansari <netchamp.faris@gmail.com>

adityahase committed:
style: Linting fixes

netchampfaris committed:
Merge branch 'staging-fixes' into prepared-report-fix
